Update runners to support A3 Mega, A3 Ultra, and A4 + add FP8 in Fuji

andersensam opened the pull request:
This PR uses the newly added runner config style to add support for A3 Mega, A3 Ultra, and A4. This also bundles NCCL configs for each platform, allowing for end users to tune themselves. FP8 training support for Fuji has also been added.

andersensam commented:
@kelvin-zou @Ethanlm Can we get some eyes on this PR to wrap up adding support for the additional accelerators? I've closed my original PR and updated the code to support the new runner config mechanism

kelvin-zou commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -919,71 +1247,255 @@ def _build_volumes(self) -> Nested[Any]:
+                # More debugging.
+                "TF_CPP_MIN_LOG_LEVEL": "0",
+                # Enable auto PGLE available in jax 0.4.33
+                "JAX_ENABLE_PGLE": "True",
```
I recommend to turn off pgle for now.

kelvin-zou commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -919,71 +1247,255 @@ def _build_volumes(self) -> Nested[Any]:
+            {
+                "LD_LIBRARY_PATH": "/usr/local/nvidia/lib64",
+                # Enable auto PGLE available in jax 0.4.33
+                "JAX_ENABLE_PGLE": "True",
```
I recommend to turn off pgle for now.

kelvin-zou commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -899,18 +987,258 @@ def _build_a3_init_container(self) -> Nested[Any]:
+                # More debugging.
+                "TF_CPP_MIN_LOG_LEVEL": "0",
+                # Enable auto PGLE available in jax 0.4.33
+                "JAX_ENABLE_PGLE": "True",
```
turn off pgle?

kelvin-zou commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -899,18 +987,258 @@ def _build_a3_init_container(self) -> Nested[Any]:
+                "JAX_REMOVE_CUSTOM_PARTITIONING_PTR_FROM_CACHE_KEY": "True",
+                # XLA team flags used this not sure what it does yet. See maxtext_wrapper.py
+                "CUDA_DEVICE_MAX_CONNECTIONS": "1",
+                "NVTE_FUSED_ATTN": "1",
```
this is not needed since we are not using TE.

kelvin-zou commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -899,18 +987,258 @@ def _build_a3_init_container(self) -> Nested[Any]:
+                # This is needed for flash attention + auto PGLE to work
+                "JAX_REMOVE_CUSTOM_PARTITIONING_PTR_FROM_CACHE_KEY": "True",
+                # XLA team flags used this not sure what it does yet. See maxtext_wrapper.py
+                "CUDA_DEVICE_MAX_CONNECTIONS": "1",
```
this seems like a MPS only feature, for training we only bind one process to one device, is it helpful here?

kelvin-zou commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -919,71 +1247,255 @@ def _build_volumes(self) -> Nested[Any]:
+                "XLA_PYTHON_CLIENT_MEM_FRACTION": "0.85",
+                "TF_FORCE_GPU_ALLOW_GROWTH": "true",
+                "NCCL_DEBUG": "WARN",
+                "NCCL_SOCKET_IFNAME": "=eth0,eth1",
```
how many interfaces do we have here? Is it only 2?

andersensam replied:
The two interfaces here are just for the control traffic -- per the internal NCCL config, this is the expected value. I've tested with it disabled and NCCL is unable to properly start

kelvin-zou replied:
I think this is something unique to your own setup. To me it looks odd to use 2 NICs, you can have 8 NICs or 1 NIC, even if you do 2 NICs for Numa purpose, I am not sure if you should have `eth0,eth1` here, can you also share your NIC topology here?

kelvin-zou commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -899,18 +987,258 @@ def _build_a3_init_container(self) -> Nested[Any]:
+                "NCCL_FASTRAK_ENABLE_CONTROL_CHANNEL": "0",
+                "NCCL_FASTRAK_ENABLE_HOTPATH_LOGGING": "0",
+                "NCCL_MIN_NCHANNELS": "4",
+                "NCCL_TUNER_PLUGIN": "libnccl-tuner.so",
```
Is it something GCP specific?

andersensam replied:
Yes, these are the required values per the GCP TCPx network setup

kelvin-zou commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -919,71 +1247,255 @@ def _build_volumes(self) -> Nested[Any]:
+                "NCCL_IB_QPS_PER_CONNECTION": "4",
+                "NCCL_IB_TC": "52",
+                "NCCL_IB_FIFO_TC": "84",
+                "NCCL_SHIMNET_GUEST_CONFIG_CHECKER_CONFIG_FILE": (
```
do we have some good documentation for this part?

kelvin-zou reviewed (commented):
Thanks, great work!

andersensam commented on axlearn/common/flash_attention/gpu_attention.py:
```diff
@@ -66,6 +66,14 @@
 
+# TODO (andersensam): Remove A4 check when Triton is updated
+# A4 platform-specific code
+import os
```
Rewrote this as a config modifier. PTAL

markblee commented on axlearn/common/flash_attention/layer.py:
```diff
@@ -341,3 +344,31 @@ def enter_fn(_, value, default_kv):
         return cfg
+
+
+class GPUFlashBlockSizeModifier(ConfigModifier):
```
Can we generalize `FlashBlockSizeModifier` to take optional `{tpu,gpu}_block_size` instead?

andersensam replied:
Good point

andersensam replied:
Consolidated in most recent commit

markblee commented on axlearn/experiments/text/gpt/fuji.py:
```diff
@@ -471,9 +476,38 @@ def get_trainer_kwargs(
+                            MeshShapeModifier.default_config().set(  
+                                mesh_shape=mesh_shape_from_axes(data=-1, fsdp=8)
+                            ),
+                            #FP8ConfigModifier.default_config().set(
```
Intended?

andersensam replied:
Indeed -- not sure that everyone wants to use FP8 training, more just showing how it should be enabled

andersensam replied:
Added the suggested comment to the newest commit -- good point and yes, it likely makes sense to create a separate config. Perhaps something like having "70B-FP8" or so

markblee commented on axlearn/experiments/trainer_config_utils.py:
```diff
@@ -39,3 +39,13 @@ class Config(FlashBlockSizeModifier.Config):
         tpu_block_size: int = 1024
+
+
+class A4FlashConfigModifier(GPUFlashBlockSizeModifier):
```
Do we need a subclass? Or can we just do:
```
config_modifiers=[
    GPUFlashBlockSizeModifier.default_config().set(gpu_block_size=64)
]
```
?

markblee reviewed (commented):
Some initial comments.

andersensam commented on axlearn/cloud/gcp/jobs/launch.py:
```diff
@@ -686,8 +686,14 @@ def _infer_runner_name(fv: flags.FlagValues = FLAGS) -> str:
-        elif fv.instance_type.startswith("gpu-a3"):
-            runner_name = "gke_gpu_a3_single"
+        elif fv.instance_type.startswith("gpu-a3-high"):
+            runner_name = "gke_gpu_a3_high"
```
Ah -- will go ahead and update these accordingly

andersensam commented on axlearn/cloud/gcp/nccl/a3_mega/guest_config.txtpb:
```diff
@@ -0,0 +1,133 @@
+# proto-file: guest_config_checker.proto
```
These are the flags GCP tests for optimal performance and our support team would ask you to revert to the provided values if you run into performance issues

andersensam commented on axlearn/cloud/gcp/runners/__init__.py:
```diff
@@ -26,17 +32,32 @@ def named_runner_configs(
             ),
             pre_provisioner=TPUNodePoolProvisioner.default_config(),
         ),
+        "gke_gpu_a3_high": GKERunnerJob.default_config().set(
```
Updating now

andersensam commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -717,77 +717,164 @@ def __call__(self) -> Sequence[Nested[Any]]:
-        """Builds a sidecar container which is required by A3
-        for GPU to GPU RDMA like networking.
+    def _build_init_containers(self) -> list[Nested[Any]]:
+        return []            
```
I think with the platforms having such different required values, having separate subclasses might make more sense, especially for readability

andersensam commented on axlearn/common/flash_attention/layer.py:
```diff
@@ -317,27 +320,33 @@ def default_output_dim_to_partition_spec(
+                if (tpu_block_size != 512):
+                    value.tpu_block_size = tpu_block_size
+                if (gpu_block_size != 128):
+                    value.gpu_block_size = gpu_block_size
```
Removed the checks. Since the default value is set, this won't change any behavior

markblee commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -717,77 +717,164 @@ def __call__(self) -> Sequence[Nested[Any]]:
+        ]
+
+        env_vars: dict[str, str] = {}
+        env_vars["DISTRIBUTED_COORDINATOR"] = f"{cfg.name}-job-0-0.{cfg.name}:8080"
```
Does `-job-` depend on `cfg.job_name`?

andersensam replied:
When using `cfg.name` we get something like `{'name': 'DISTRIBUTED_COORDINATOR', 'value': 'samuelandersen-axlearn-syjmle-job-0-0.samuelandersen-axlearn-syjmle:8080'}`

With `cfg.job_name` with `f"{cfg.job_name}-0-0.{cfg.name}:8080"` we just get: `{'name': 'DISTRIBUTED_COORDINATOR', 'value': 'job-0-0.samuelandersen-axlearn-syjmle:8080'}`

Ethanlm replied:
The "-job-" string in `f"{cfg.name}-job-0-0.{cfg.name}:8080"` actually depends on how `cfg.job_name`

https://github.com/apple/axlearn/blob/682a718fd19253e74454ddd125ed6cfed9bb4329/axlearn/cloud/gcp/jobset_utils.py#L231

`cfg.job_name` is default to "job". If it is set to something else, for example, "XYZ", then the coordinator will be 
`f"{cfg.name}-XYZ-0-0.{cfg.name}:8080"`

markblee replied:
Seems like this hasn't been addressed?

andersensam replied:
Updated in the latest commit -- changed to `env_vars["DISTRIBUTED_COORDINATOR"] = f"{cfg.name}-{cfg.job_name}-0-0.{cfg.name}:8080"`

markblee commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -717,77 +717,164 @@ def __call__(self) -> Sequence[Nested[Any]]:
 
-        volume_mounts = [
-            {"name": "shared-memory", "mountPath": "/dev/shm"},
+        GPUReplicatedJob_main_container: Nested[Any] = super()._build_main_container()
```
```suggestion
        base_main_container: Nested[Any] = super()._build_main_container()
```
It's uncommon to include class names in variable names.

markblee commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -899,18 +987,238 @@ def _build_a3_init_container(self) -> Nested[Any]:
+        command = [
+            "bash",
+            "-c",
+            'set -x; /tcpgpudmarxd/build/app/tcpgpudmarxd --gpu_nic_preset a3vm  \
+                --gpu_shmem_type fd --uds_path /run/tcpx \
+                --setup_param "--verbose 128 2 0" & \n\
+            while [ ! -f /run/tcpx/terminated ]; do sleep 10; done;',
```
Link to https://cloud.google.com/compute/docs/gpus/gpudirect#provide-access for ref?

andersensam replied:
Added

markblee commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -899,18 +987,238 @@ def _build_a3_init_container(self) -> Nested[Any]:
+
+    def _build_init_containers(self) -> Nested[Any]:
+        return [self._build_a3_high_tcpx_init_container(), self._build_a3_high_sidecar_container()]
+
```
```suggestion


```

markblee commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -899,18 +987,238 @@ def _build_a3_init_container(self) -> Nested[Any]:
+        command = [
+            "bash",
+            "-c",
+            'set -ex; chmod 755 /scripts/container_entry.sh; \n\
+             /scripts/container_entry.sh install; \n\
+             mkdir -p /usr/lib/tcpx/lib64; \n\
+             cp -r /var/lib/tcpxo/lib64/. /usr/lib/tcpx/lib64; \n\
+             echo "installation finishes";',
```
Likewise add a ref here and below?

andersensam replied:
Added

markblee replied:
Where did you add this?

andersensam replied:
Had it on line 1098, but made a copy of the comment for readability in the init container too -- thanks

markblee commented on axlearn/cloud/gcp/runners/__init__.py:
```diff
@@ -39,9 +40,34 @@ def named_runner_configs(
     raise ValueError(f"Unrecognized runner: {runner}")
+
+
+def get_gpu_runners() -> dict[str, BaseRunnerJob.Config]:
```
```suggestion
def _get_gpu_runners() -> dict[str, BaseRunnerJob.Config]:
```

markblee commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -899,18 +987,238 @@ def _build_a3_init_container(self) -> Nested[Any]:
+
+        env_vars: dict[str, str] = GPUReplicatedJob_main_container["env"]
+
+        default_xla_flags = [
```
Anytime we set a default, we try to also add references (or comments, if they were manually tuned) so that they can be updated as necessary. 

(This also holds for the many env vars that we're setting below.)

andersensam replied:
Added some comments / references for the vars

markblee commented on axlearn/common/flash_attention/layer.py:
```diff
@@ -48,8 +48,10 @@ class Config(GroupedQueryAttention.Config):
         causal: bool = False
         # The block size used to tile attention computation (for TPU only).
         # Should be less than the target sequence length and a multiple of 128 on TPU.
-        # TODO(tom_gunter): Expose GPU block-size (currently always 128) & unify.
```
I think we'd still want to unify tpu/gpu_block_size as they are never used concurrently. Can you retain but update the TODO?

andersensam replied:
Returned the TODO and accepted your proposed changes + validated with the config test

markblee commented on axlearn/common/flash_attention/layer.py:
```diff
@@ -48,8 +48,10 @@ class Config(GroupedQueryAttention.Config):
         tpu_block_size: int = 512
+        # The default GPU block-size of 128 works on most accelerators
+        # NVIDIA Blackwell (B200) requires a smaller block-size
+        gpu_block_size: int = 128
```
```suggestion
        gpu_block_size: Optional[int] = None
```
To avoid many golden config changes.

markblee commented on axlearn/common/flash_attention/layer.py:
```diff
@@ -195,6 +197,7 @@ def _compute_attention(
             is_decoding=is_decoding,
             # TODO(hanzhi-zhou): Refactor backend specific config passing.
             tpu_block_size=cfg.tpu_block_size,
+            gpu_block_size=cfg.gpu_block_size,
```
```suggestion
            gpu_block_size=cfg.gpu_block_size or 128,
```

markblee reviewed (commented):
Thanks @andersensam -- one note is that we usually ask contributors not to resolve comments, but instead let reviewers do so: https://github.com/apple/axlearn/blob/main/CONTRIBUTING.md#code-review-protocol

kelvin-zou commented on axlearn/common/flash_attention/layer.py:
```diff
@@ -48,8 +48,10 @@ class Config(GroupedQueryAttention.Config):
-        # TODO(tom_gunter): Expose GPU block-size (currently always 128) & unify.
         tpu_block_size: int = 512
+        # The default GPU block-size of 128 works on most accelerators
+        # NVIDIA Blackwell (B200) requires a smaller block-size
```
is this comment accurate? Why B200 is asking for a smaller block size?

andersensam replied:
B200 requires a smaller block size for some of the backwards kernel operations based on architectural changes and memory layouts. Triton / Pallas are aware of this and there should be a fix in the works. This value is only used when launching the backwards Pallas kernel in `gpu_attention.py` and does not affect CuDNN or other kernels

kelvin-zou replied:
Cool, good to know! Sgtm!

andersensam commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -899,18 +987,238 @@ def _build_a3_init_container(self) -> Nested[Any]:
+        env_vars: dict[str, str] = GPUReplicatedJob_main_container["env"]
+
+        default_xla_flags = [
+            "--xla_gpu_enable_latency_hiding_scheduler=true",
```
Only one of the flags is shared and many are changed depending on the platform. I think it makes sense to have them explicitly listed per platform to keep the configs in one block, but open to changing if desired

kelvin-zou replied:
Hmm, I think the only flags changed are for the combiner thresholds, the rest look the same to me (with different orders). Can you actually group them into `GLOBAL_GPU_XLA_FLAGS` and local variables like `H100_XLA_FLAGS`, `B200_XLA_FLAGS` etc? It would make it much easier to read and revise from our end.

kelvin-zou reviewed (commented):
Mostly minor comments about jobset_util GPU related flags, overall change lgtm.

andersensam commented:
Moved XLA flags to a `global_gpu_xla_flags` variable in `GPUReplicatedJob` per @kelvin-zou. Also removed extraneous NCCL config: I tested removing `NCCL_SOCKET_IFNAME` for IB networks and it has no effect. Removing it allows for the auto-discovery and shows no issues when running multi-node testing this morning.

Ethanlm commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -717,80 +717,183 @@ def __call__(self) -> Sequence[Nested[Any]]:
+        cfg: GPUReplicatedJob.Config = self.config
+        volumes = self._build_volumes()
+        annotations = {
+            "kubectl.kubernetes.io/default-container": cfg.name,
```
I understand this was also there in the original code. But curious if you know what this annotation is for?

https://kubernetes.io/docs/reference/labels-annotations-taints/#kubectl-kubernetes-io-default-container
Does it have any other implication besides annotating?

samos123 replied:
It's helpful when you run things like `kubectl logs` and there are sidecar containers. We can probably get rid of it now that sidecar containers are part of initContainers.

kelvin-zou reviewed (approved):
Thanks! Lgtm from GPU config side.
Also cc @matthew-e-hopkins  @hanzhi713  for your awareness about the B200/H200 configs.

markblee commented on axlearn/cloud/gcp/jobset_utils.py:
```diff
@@ -717,80 +717,183 @@ def __call__(self) -> Sequence[Nested[Any]]:
+            "--xla_gpu_enable_all_gather_combine_by_dim=false",
+            "--xla_gpu_enable_reduce_scatter_combine_by_dim=false",
+            "--xla_disable_hlo_passes=rematerialization",
+            " ", # Leave a trailing space for A3 / A4-specific XLA flags to be added
```
What does this mean? Shouldn't the flags be space-agnostic?

andersensam replied:
Since I'm doing a string concatenation inside each class, I thought it might be cleaner to have a space at the end of `env["XLA_FLAGS"]`, so then I can use the `+=` operator after `" ".join(platform_xla_flags)`

markblee reviewed (commented):
Thanks @andersensam ! Structurally the PR looks fine to me. We'll need to run some internal tests before merging.

andersensam commented on axlearn/experiments/trainer_config_utils.py:
```diff
@@ -29,13 +28,3 @@ def wrapped_fn():
         return trainer_cfg
 
     return wrapped_fn
-
```
Added back with relevant tests

markblee commented on axlearn/common/flash_attention/layer.py:
```diff
@@ -317,16 +321,18 @@ def default_output_dim_to_partition_spec(
 
-        tpu_block_size: Required[int] = REQUIRED
+        tpu_block_size: Optional[int] = 512
+        gpu_block_size: Optional[int] = 128
```
```suggestion
        gpu_block_size: Optional[int] = None
```
As otherwise it incurs golden config changes across various configs. Since we've already set a default, we probably don't need to repeat it here.

markblee commented:
If you rebase to `main`, I think the CI failure should go away.

Ethanlm commented:
@muyangyuapple has an internal PR that will run with this PR into internal CI. 
Let's also wait for that before we can merge. Thanks!

andersensam commented on axlearn/cloud/gcp/job_test.py:
```diff
@@ -132,7 +132,9 @@ def _job_config(
         fv = flags.FlagValues()
-        cfg = job.GKEJob.default_config().set(builder=jobset_utils.A3ReplicatedJob.default_config())
+        cfg = job.GKEJob.default_config().set(
+            builder=jobset_utils.A3HighReplicatedJob.default_config()
```
We use `A3HighReplicatedJob`, though it really could be that or any variety of A3 / A4. `GPUReplicatedJob` is just a generic class that won't produce a complete `JobSpec` to validate in the testing

Ethanlm commented:
Internal CI check has passed. I think we can merge once @markblee approves
